### PR TITLE
Sampler that splits utterances into buckets

### PR DESCRIPTION
**Pull request by EgorLakomkin**
I think it might be a solution to #30. Instead of random sampling, we could split firstly the whole dataset into several buckets depending on the length of the utterance and then randomly sample from each of the buckets. As a result, we can have batches with similar length utterances.

### CODE REVIEW

**Comment by SeanNaren** [comment]
Thanks for this @EgorLakomkin, has this shown an improvement of memory usage when profiled or checking nvidia-smi once this kicks in?

**Comment by EgorLakomkin** [comment]
I am also testing on librispeech+ted+voxforge for a couple of epochs to measure if memory blows up in random sampling.

**Comment by ryanleary** [comment]
I've run a few epochs of librispeech 1k hrs training with this branch and still ended up with a random OOM.

I think this is a step in the right direction, but I wonder if we can't take it a step further to try to guarantee a maximum sequence length.

**Comment by EgorLakomkin** [comment]
Strange, mine is working quit well.
Did you limit the length of the utterances? I have tried with max 15 seconds.

**Comment by ryanleary** [comment]
Nope. Not limiting the length anywhere (where are you doing that, by the way?). I think the longest utterance in the LibriSpeech corpus is something like 35 seconds.

**Comment by EgorLakomkin** [comment]
There is an option in merge_manifests script to limit minimum or maximum length. 35 seconds is indeed long, I guess for such long utterances the only solution would be to decrease that batch size.

**Comment by EgorLakomkin** [comment]
I have tried several times and also caught OOM. Set-up is batch_size 16, gru 5 layer, 1024 hidden size, max length 15.

Now I have added one more thing: del loss & del out references during training. I have seen somewhere that it might help with OOMs. The same set-up is working for me now without OOMs

**Comment by ryanleary** [comment]
Also saw that in the forums recently. Glad that's helping.

**Comment by SeanNaren** [comment]
Thought I'd help out, finished the review and dealt with merge commits on the main branch. Let me know what you think!

File: train.py [hunk 1]
```diff
@@ -52,6 +53,7 @@
 parser.add_argument('--tensorboard', dest='tensorboard', action='store_true', help='Turn on tensorboard graphing')
 parser.add_argument('--log_dir', default='visualize/deepspeech_final', help='Location of tensorboard log')
 parser.add_argument('--log_params', dest='log_params', action='store_true', help='Log parameter values and gradients')
+parser.add_argument('--bucketing', dest='bucketing', action='store_true', help='Split utterances into buckets and sample from them')
```
Comment (ryanleary) [inline train.py]: Concur

**Comment by EgorLakomkin** [comment]
Looks good! Thank you :)

File: data/bucketing_sampler.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+        audio_paths = [path for (path, _) in self.ids]
+        audio_lengths = [len(load_audio(path)) for path in audio_paths ]
+        hist, bin_edges = np.histogram(audio_lengths, bins="auto")
+        audio_samples_indices = np.digitize(audio_lengths, bins=bin_edges)
+        self.bins_to_samples = defaultdict(list)
+        for idx, bin_id in enumerate(audio_samples_indices):
+            self.bins_to_samples[bin_id].append(idx)
+
+class BucketingSampler(Sampler):
+    """
```
Comment (EgorLakomkin) [inline data/bucketing_sampler.py]: I have added few comments on SpectrogramDatasetWithLength

**Comment by SeanNaren** [comment]
Pulling in this now and thanks for the `del` fix, seems like its been doing good things for memory!

**Pull request merged**